Challenge-4/Bot_2/src/agent.spec.ts

```diff
@@ -0,0 +1,459 @@
+    `function getSupplyRate(uint utilization) public view returns (uint64)`,
+    `function getBorrowRate(uint utilization) public view returns (uint64)`,
+  ];
+  const OTHER_FUNCTION = "function otherFunction(address asset, uint amount)";
+  const Iface = new ethers.utils.Interface(functions);
+  const provideInterface = new ethers.utils.Interface([
+    "function supply(address asset, uint amount)",
+    "function withdraw(address asset, uint amount)",
+    OTHER_FUNCTION,
+  ]);
+
+  const mockAlerts = (alertId: string, kink: any, utilization: any) => {
+    (getAlerts as jest.Mock).mockResolvedValue({
+      alerts: [
+        {
+          alertId: alertId,
+          hasAddress: jest.fn().mockReturnValue(false),
+          metadata: {
+            kink: kink,
+            utilization: utilization,
+          },
+        },
+      ],
+      pageInfo: { hasNextPage: false },
+    });
+  };
+
+  beforeEach(() => {
+    handleTransaction = provideHandleTransaction();
+    initialize = provideInitialize(mockProvider as any);
+    helper = new Helper(mockProvider as any);
+    txEvent = new TestTransactionEvent().setBlock(0);
+  });
+
+  let setupMockProvider = async (
+    supplyKink: ethers.BigNumber,
+    borrowKink: ethers.BigNumber,
+    utilization: ethers.BigNumber,
+  ) => {
+    mockProvider.setNetwork(1);
+    mockProvider.addCallTo(
+      mockConfiguratorProxy,
+      0,
+      Iface,
+      "getConfiguration",
+      {
+        inputs: [usdcTokenAddress],
+        outputs: [
+          {
+            governor: createAddress("0x123"),
+            pauseGuardian: createAddress("0x2375"),
+            baseToken: createAddress("0x98235"),
+            baseTokenPriceFeed: createAddress("0x1274"),
+            extensionDelegate: createAddress("0x5124"),
+            supplyKink: supplyKink,
+            supplyPerYearInterestRateSlopeLow: 60,
+            supplyPerYearInterestRateSlopeHigh: 100,
+            supplyPerYearInterestRateBase: 900,
+            borrowKink: borrowKink,
+            borrowPerYearInterestRateSlopeLow: 100,
+            borrowPerYearInterestRateSlopeHigh: 200,
+            borrowPerYearInterestRateBase: 300,
+            storeFrontPriceFactor: 50,
+            trackingIndexScale: 50,
+            baseTrackingSupplySpeed: 50,
+            baseTrackingBorrowSpeed: 50,
+            baseMinForRewards: 50,
+            baseBorrowMin: 50,
+            targetReserves: 50,
+            assetConfigs: [
+              {
+                asset: usdcTokenAddress,
+                decimals: 18,
+                conversionFactor: 1,
+              },
+            ],
+          },
+        ],
+      },
+    );
+    mockProvider.addCallTo(usdcTokenAddress, 0, Iface, "getUtilization", {
+      inputs: [],
+      outputs: [utilization],
+    });
+    mockProvider.addCallTo(usdcTokenAddress, 0, Iface, "getSupplyRate", {
+      inputs: [utilization],
+      outputs: [986453221],
```

Comment:
This, and the output for `getBorrowRate` could be declared globally then passed into this array to make it easier to read. Plus, if you reuse the value, it's easier if you only have to change them in one location.